python/cudf/cudf/core/frame.py

```diff
@@ -438,128 +438,241 @@ def _normalize_columns_and_scalars_type(self, other):
-        0    4
-        1    3
-        2
-        3
-        4
+        dtype: int64
+        >>> ser.where(ser > 2)
+        0       4
+        1       3
+        2    null
+        3    null
+        4    null
+        dtype: int64
         """
 
         if isinstance(self, cudf.DataFrame):
-            if not isinstance(boolean_mask, cudf.DataFrame):
-                boolean_mask = self.from_pandas(pd.DataFrame(boolean_mask))
+            if not isinstance(cond, cudf.DataFrame):
+                cond = self.from_pandas(pd.DataFrame(cond))
+
+            common_cols = set(self.columns).intersection(set(cond.columns))
+            if len(common_cols) > 0:
+                # If `self` and `cond` are having unequal index,
+                # then re-index `cond`.
+                if len(self.index) != len(cond.index) or any(
+                    self.index != cond.index
+                ):
+                    cond = cond.reindex(self.index)
+            else:
+                if cond.shape != self.shape:
+                    raise ValueError(
+                        """Array conditional must be same shape as self"""
+                    )
+                # Setting `self` column names to `cond`
+                # as `cond` has no column names.
+                cond.columns = self.columns
+
             other = self._normalize_columns_and_scalars_type(other)
             out_df = cudf.DataFrame(index=self.index)
             if len(self._columns) != len(other):
                 raise ValueError(
                     """Replacement list length or number of dataframe columns
                     should be equal to Number of columns of dataframe"""
                 )
-            if len(self._columns) != len(boolean_mask._columns):
-                raise ValueError(
-                    """Array conditional must be same shape as self"""
-                )
 
-            for in_col_name, cond_col_name, otr_col in zip(
-                self.columns, boolean_mask.columns, other
-            ):
-                input_col = self[in_col_name]._column
-                if is_categorical_dtype(input_col.dtype):
-                    if np.isscalar(otr_col):
-                        otr_col = input_col._encode(otr_col)
-                    else:
-                        otr_col = otr_col.codes
-                    input_col = input_col.codes
+            for column_name, other_column in zip(self.columns, other):
+                input_col = self._data[column_name]
+                if column_name in cond.columns:
+                    if is_categorical_dtype(input_col.dtype):
+                        if np.isscalar(other_column):
+                            try:
+                                other_column = input_col._encode(other_column)
+                            except ValueError:
+                                # When other is not present in categories.
+                                other_column = input_col.default_na_value()
```

Comment:
Shouldn't this just be null instead of `default_na_value()`?

Reply:
So this was done to match pandas behaviour where their codes are -1 if there is a value which is not in category is being passed in `other_column`.

Reply:
We generally don't do that IIRC, we set the value to null.

Reply:
Changed the logic to set it to `null`.